### DISCUSSION TITLE
사용자 수준 스레드와 커널 수준 스레드의 차이를 설명해주세요. 7/13

### DISCUSSION THREAD

**Discussion by SionBackEnd** [Team-Psyduck/tech-sharing#70] [General]
오늘의 알고리즘 
[프로그래머스 소수찾기](https://school.programmers.co.kr/learn/courses/30/lessons/42839)

**Comment by SionBackEnd** [comment]
사용자 수준 스레드는 프로세스 내부 스레드로 운영체제가 알지 못하는 스레드.
- 사용자 수준 스레드 N개가 커널 수준 스레드 1개에 매핑되므로, 다대일 스레드 매핑이라고 함. 
- 스레드 동기화를 위해 커널을 호출하지 않아 오버헤드가 적다는 장점을 가짐.
- 하지만, 사용자 수준 스레드 중에서 하나라도 block 상태가 된다면 모든 스레드가 block 상태가 됨.

커널 수준 스레드는 운영체제가 관리하는 스레드. 
- 사용자 수준 스레드와 커널 수준 스레드가 1대1로 매핑.
- 사용자 수준 스레드의 존재를 알지 못함.
- 사용자 수준 스레드와 다르게 하나의 스레드가 block상태가 되어도 다른 스레드들은 block상태가 되지 않고 독립적으로 일을 할 수 있음. 
- 이때 커널 수준 스레드는 다른 스레드를 CPU자리에 놓기 위해서 컨텍스트 스위칭이라는 작업을 진행하여 PC데이터를 교체하고 캐시데이터를 비우는 작업을 진행하여 오버헤드가 커짐.

[참고 사이트](https://helloinyong.tistory.com/293)

**Comment by eatingbug** [comment]
- Kernel Level Thread 와 User Level Thread 는 생성 주체가 누구냐에 따라 구분된다.
- 프로그래머의 요청에 따라 Thread 를 생성하고 Scheduling 하는 주체가 Kernel 이면 `Kernel Level Thread` 라고 한다.
- Kernel 이 Thread Model 을 지원하지 않거나 제공하는 Thread Model 이 마음에 들지 않을 경우, Kernel 에 의존적이지 않은 형태로 Thread 의 기능을 제공하는 라이브러리를 활용할 수 있는데, 이러한 방식으로 제공되는 Thread 를 `User Level Thread` 라고 한다.

출처 : [사용자 수준 Thread와 커널 수준 Thread 의 차이점은?](https://junghyun100.github.io/%EC%82%AC%EC%9A%A9%EC%9E%90%EC%88%98%EC%A4%80ThreadVS%EC%BB%A4%EB%84%90%EC%88%98%EC%A4%80Thread/)
